### ISSUE TITLE
SmolVLA: partial-freeze in set_requires_grad() never matches — last VLM layer + final norm are silently left trainable when train_expert_only=False

### ISSUE THREAD

**Issue by zuoxingdong** [huggingface/lerobot#4018]
### Ticket Type

🐛 Bug Report (Something isn't working)

### Environment & System Info

```Shell

```

## 🐛 Summary

In `src/lerobot/policies/smolvla/smolvlm_with_expert.py`, the `train_expert_only=False`
branch of `set_requires_grad()` is supposed to freeze the **last VLM layer + final norm +
lm_head** (per the code comment, "to avoid unused params issue with distributed training").

The name patterns it matches against **do not exist in the model**, so the freeze silently
no-ops. Only `lm_head` gets frozen (by substring luck). This has been the case **since
SmolVLA was first merged (#1175)** — it is not caused by a transformers upgrade; the
patterns never matched on any supported transformers version (verified on 4.50.3, the
original pin, and 5.5.4).

## 🔍 Where

`smolvlm_with_expert.py`, `set_requires_grad()`:

```python
frozen_layers = [
    "lm_head",
    "text_model.model.norm.weight",          # ❌ no such name
]
for layer in last_layers:
    frozen_layers.append(f"text_model.model.layers.{layer}.")   # ❌ no such name

for name, params in self.vlm.named_parameters():
    if any(k in name for k in frozen_layers):
        params.requires_grad = False
```

| Pattern searched | Real parameter name | Matches? |
|---|---|---|
| `text_model.model.norm.weight` | `model.text_model.norm.weight` | ❌ never |
| `text_model.model.layers.15.` | `model.text_model.layers.15.` | ❌ never |
| `lm_head` | `model.lm_head.weight` | ✅ (substring) |

There is no `.model` nesting inside `text_model`: `SmolVLMModel.text_model` is built with
`AutoModel.from_config(...)` → a bare `LlamaModel` with `.layers` directly. The nested
pattern appears to be carried over from pi0's `paligemma_with_expert.py`, where
`language_model.model.layers` **is** correct (`language_model` there is a ForCausalLM
wrapper).

## 💥 Impact

Only the **VLM-finetuning** path (`train_expert_only=False`) is affected — the default
config (`train_expert_only=True`) freezes the whole VLM and bypasses this branch, which is
probably why it went unnoticed.

| Parameter group | Intended | Actual |
|---|---|---|
| `lm_head` (~47M) | ❄️ frozen | ❄️ frozen |
| last VLM layer (~9.8M with default `num_vlm_layers=16`) | ❄️ frozen | 🔥 trainable |
| final norm (`text_model.norm.weight`) | ❄️ frozen | 🔥 trainable |

Consequences:

1. **DDP failures** 💣 — the final norm's output (final prefix hidden state) is discarded,
   so it is trainable-but-gradient-less → the exact "unused parameters" DDP error this
   freeze was written to prevent.
2. **Silent training deviation** — parameters that the recipe says are frozen (the last
   VLM layer's k/v projections) actually receive gradients through the joint-attention KV
   path and train.
3. **Misleading trainable-parameter counts** (~10M inflated).

## 🧪 Reproduction

```python
from transformers import AutoConfig
from transformers.models.smolvlm import SmolVLMForConditionalGeneration

cfg = AutoConfig.from_pretrained("HuggingFaceTB/SmolVLM2-500M-Video-Instruct")
vlm = SmolVLMForConditionalGeneration(config=cfg)

print(sum("text_model.model." in n for n, _ in vlm.named_parameters()))  # 0 😬
print([n for n, _ in vlm.named_parameters() if "text_model.norm" in n])
# ['model.text_model.norm.weight']
```

Measured freeze coverage (full 32-layer SmolVLM2-500M, identical result on
transformers 4.50.3 and 5.5.4):

| Matcher | Tensors frozen | Params frozen |
|---|---|---|
| current (buggy) | 1 (lm_head only) | 47.3M |
| corrected | 20 | 67.0M |

## ✅ Suggested fix

Use the real (flat) names, and fail loudly if the patterns ever stop matching — a silent
no-op is how this bug survived unnoticed:

```diff
             frozen_layers = [
                 "lm_head",
-                "text_model.model.norm.weight",
+                "text_model.norm.weight",
             ]
             for layer in last_layers:
-                frozen_layers.append(f"text_model.model.layers.{layer}.")
+                frozen_layers.append(f"text_model.layers.{layer}.")

+            n_frozen = 0
             for name, params in self.vlm.named_parameters():
                 if any(k in name for k in frozen_layers):
                     params.requires_grad = False
+                    n_frozen += 1
+            assert n_frozen > 0, (
+                f"Freeze patterns matched no parameters — transformers layout changed? {frozen_layers}"
+            )
```

PR: #4019 

## 🖥️ Environment

- lerobot: current `main` (bug present since #1175)
- transformers: verified on 4.50.3 and 5.5.4 (the whole `transformers>=5.4.0,<5.6.0` pin range is affected)
- Trigger: `policy.train_expert_only=false` + multi-GPU (DDP) training

### Relevant logs or stack trace

```Shell

```

### Checklist

- [x] I have searched existing tickets to ensure this isn't a duplicate.
- [x] I am using the latest version of the `main` branch.
- [x] I have verified this is not an environment-specific problem.
